mains/gdas.cc

```diff
@@ -0,0 +1,122 @@
+// -------------------------------------------------------------------------------------------------
+
+#include <functional>
+#include <map>
+
+#include "fv3jedi/ObsLocalization/instantiateObsLocFactory.h"
+#include "fv3jedi/Utilities/Traits.h"
+
+#include "soca/Traits.h"
+
+#include "oops/generic/instantiateModelFactory.h"
+#include "saber/oops/instantiateCovarFactory.h"
+#include "ufo/instantiateObsErrorFactory.h"
+#include "ufo/instantiateObsFilterFactory.h"
+#include "ufo/ObsTraits.h"
+
+#include "oops/runs/ConvertState.h"
+#include "oops/runs/HofX4D.h"
+#include "oops/runs/LocalEnsembleDA.h"
+#include "oops/runs/Run.h"
+#include "oops/runs/Variational.h"
+
+// -------------------------------------------------------------------------------------------------
+
+template<typename Traits>
+int runApp(int argc, char** argv, const std::string traits, const std::string appName) {
+  // Create the Run object
+  oops::Run run(argc, argv);
+
+  // Instantiate oops factories
+  oops::instantiateModelFactory<Traits>();
+
+  // Instantiate saber factories
+  saber::instantiateCovarFactory<Traits>();
+
+  // Intantiate ufo factories
+  ufo::instantiateObsErrorFactory();
+  ufo::instantiateObsFilterFactory();
+
+  // Localization for ensemble DA
+  if (appName == "localensembleda") {
+    if (traits == "fv3jedi") {
+      fv3jedi::instantiateObsLocFactory();
+    } else if (traits == "soca") {
+      ufo::instantiateObsLocFactory<soca::Traits>();
+    }
+  }
+
+  // Application pointer
+  std::unique_ptr<oops::Application> app;
+
+  // Define a map from app names to lambda functions that create unique_ptr to Applications
+  std::map<std::string, std::function<std::unique_ptr<oops::Application>()>> apps;
+
+  apps["convertstate"] = []() {
```

Comment:
do we need a ConvertToStructuredGrid?

Reply:
Also the fv3jedi error covariance toolbox?

Reply:
I can add those. Do you know if these are used anywhere or just for future use? I didn't notice them being used in my grepping, but perhaps I missed a section.

Reply:
@andytangborn has definitely been using the error covariance toolbox and I think we will need the ConvertToStructuredGrid for GFSv17 for aerosol products (it's in a branch, just haven't been able to put it in the workflow develop yet)

Reply:
I added fv3jedi error covariance toolbox (testing that it builds now) but I think JEDI in GDAS doesn't yet have the relevant commits for ConvertToStructuredGrid so we can add that one as the need arises.

Reply:
Ok let's merge as is and when the hashes are updated, we can add that in then, it seems like it should be easy to extend now that you have an example here.